crates/holochain/src/core/workflow/app_validation_workflow/unit_tests.rs

```diff
@@ -0,0 +1,506 @@
+    test_space.space.cache_db.test_write(move |txn| {
+        insert_op(txn, &dht_op_hashed).unwrap();
+    });
+
+    // the same validation should now successfully validate the op
+    let outcome = run_validation_callback_inner(
+        invocation.clone(),
+        &ribosome,
+        workspace_read.clone(),
+        network.clone(),
+        fetched_dependencies.clone(),
+    )
+    .await
+    .unwrap();
+    assert_matches!(outcome, Outcome::Accepted);
+}
+
+// test that unresolved dependent hashes are fetched
+#[tokio::test(flavor = "multi_thread")]
+async fn validation_callback_awaiting_deps_hashes() {
+    holochain_trace::test_run().unwrap();
+
+    let zomes = SweetInlineZomes::new(vec![], 0).integrity_function("validate", {
+        move |api, op: Op| {
+            if let Op::RegisterAgentActivity(RegisterAgentActivity { action, .. }) = op {
+                let result = api.must_get_action(MustGetActionInput(action.as_hash().to_owned()));
+                if result.is_ok() {
+                    Ok(ValidateCallbackResult::Valid)
+                } else {
+                    Ok(ValidateCallbackResult::UnresolvedDependencies(
+                        UnresolvedDependencies::Hashes(vec![action.as_hash().clone().into()]),
+                    ))
+                }
+            } else {
+                Ok(ValidateCallbackResult::Valid)
+            }
+        }
+    });
+
+    let TestCase {
+        agent_key,
+        dna_file,
+        integrity_zomes,
+        ribosome,
+        workspace_read,
+        ..
+    } = TestCase::new(zomes).await;
+
+    let action = fixt!(Action);
+    let action_signed_hashed = SignedActionHashed::new_unchecked(action.clone(), fixt!(Signature));
+    let action_op = Op::RegisterAgentActivity(RegisterAgentActivity {
+        action: action_signed_hashed.clone(),
+        cached_entry: None,
+    });
+
+    let zomes_to_invoke = ZomesToInvoke::OneIntegrity(integrity_zomes[0].clone());
+    let invocation = ValidateInvocation::new(zomes_to_invoke, &action_op).unwrap();
+    let dna_hash = dna_file.dna_hash().clone();
+
+    // handle only Get events
+    let filter_events = |evt: &_| match evt {
+        holochain_p2p::event::HolochainP2pEvent::Get { .. } => true,
+        _ => false,
+    };
+    let (tx, mut rx) = tokio::sync::mpsc::channel(1);
+    let network = test_network_with_events(
+        Some(dna_hash.clone()),
+        Some(agent_key.clone()),
+        filter_events,
+        tx,
+    )
+    .await;
+
+    // respond to Get request with requested action
+    tokio::spawn({
+        let action_hash = action.clone().to_hash();
+        let action_hash_32 = action_hash.get_raw_32().to_vec();
+        async move {
+            while let Some(evt) = rx.recv().await {
+                if let HolochainP2pEvent::Get {
+                    dht_hash, respond, ..
+                } = evt
+                {
+                    assert_eq!(dht_hash.get_raw_32().to_vec(), action_hash_32);
```

Comment:
Is there a shorter way of comparing an ActionHash with an AnyDhtHash?

Reply:
There should be an `into`, I would expect something like 

`assert_eq!(dht_hash, action_hash.into());`

to work

Reply:
```rust
assert_eq!(dht_hash, AnyDhtHash::from(action_hash));
```

https://github.com/holochain/holochain/blob/83efb693ce64f0bd628e11da08eece8a743b2633/crates/holo_hash/src/aliases.rs#L209